From 9b38ea8a0c69ab9ffae78449c86265f6de66c393 Mon Sep 17 00:00:00 2001
From: Joey Dumont <joey.dumont@gmail.com>
Date: Sun, 2 May 2021 06:37:03 -0400
Subject: [PATCH] Fix -Wformat-security warnings in libcpp/{expr.c,macro.c}.

---
 libcpp/expr.c  | 19 +++++++++++--------
 libcpp/macro.c |  8 ++++----
 2 files changed, 15 insertions(+), 12 deletions(-)

diff --git a/libcpp/expr.c b/libcpp/expr.c
index dd5611dce..27f3f83af 100644
--- a/libcpp/expr.c
+++ b/libcpp/expr.c
@@ -809,10 +809,10 @@ cpp_classify_number (cpp_reader *pfile, const cpp_token *token,
 
 	  if (CPP_OPTION (pfile, c99))
             cpp_warning_with_line (pfile, CPP_W_LONG_LONG, virtual_location,
-				   0, message);
+				   0, "%s", message);
           else
             cpp_pedwarning_with_line (pfile, CPP_W_LONG_LONG,
-				      virtual_location, 0, message);
+				      virtual_location, 0, "%s", message);
         }
 
       if ((result & CPP_N_SIZE_T) == CPP_N_SIZE_T
@@ -822,7 +822,7 @@ cpp_classify_number (cpp_reader *pfile, const cpp_token *token,
 				? N_("use of C++23 %<size_t%> integer constant")
 				: N_("use of C++23 %<make_signed_t<size_t>%> integer constant");
 	  cpp_warning_with_line (pfile, CPP_W_SIZE_T_LITERALS,
-				 virtual_location, 0, message);
+				 virtual_location, 0, "%s", message);
        }
 
       result |= CPP_N_INTEGER;
@@ -836,12 +836,15 @@ cpp_classify_number (cpp_reader *pfile, const cpp_token *token,
     {
       if (!CPP_OPTION (pfile, binary_constants)
 	  && CPP_PEDANTIC (pfile))
+      {
+	const char* message = CPP_OPTION(pfile, cplusplus)
+	                      ? N_("binary constants are a C++14 feature "
+				   "or GCC extension")
+	                      : N_("binary constants are a C2X feature "
+				  "or GCC extension");
 	cpp_error_with_line (pfile, CPP_DL_PEDWARN, virtual_location, 0,
-			     CPP_OPTION (pfile, cplusplus)
-			     ? N_("binary constants are a C++14 feature "
-				  "or GCC extension")
-			     : N_("binary constants are a C2X feature "
-				  "or GCC extension"));
+			     "%s", message);
+      }
       else if (CPP_OPTION (pfile, cpp_warn_c11_c2x_compat) > 0)
 	cpp_warning_with_line (pfile, CPP_W_C11_C2X_COMPAT,
 			       virtual_location, 0,
diff --git a/libcpp/macro.c b/libcpp/macro.c
index dff7c98a4..5c7a472dd 100644
--- a/libcpp/macro.c
+++ b/libcpp/macro.c
@@ -181,7 +181,7 @@ class vaopt_state {
 	if (m_state == 2 && token->type == CPP_PASTE)
 	  {
 	    cpp_error_at (m_pfile, CPP_DL_ERROR, token->src_loc,
-			  vaopt_paste_error);
+			  "%s", vaopt_paste_error);
 	    return ERROR;
 	  }
 	/* Advance states before further considering this token, in
@@ -210,7 +210,7 @@ class vaopt_state {
 		if (was_paste)
 		  {
 		    cpp_error_at (m_pfile, CPP_DL_ERROR, token->src_loc,
-				  vaopt_paste_error);
+				  "%s", vaopt_paste_error);
 		    return ERROR;
 		  }
 
@@ -3602,7 +3602,7 @@ create_iso_definition (cpp_reader *pfile)
 	     function-like macros, but not at the end.  */
 	  if (following_paste_op)
 	    {
-	      cpp_error (pfile, CPP_DL_ERROR, paste_op_error_msg);
+	      cpp_error (pfile, CPP_DL_ERROR, "%s", paste_op_error_msg);
 	      goto out;
 	    }
 	  if (!vaopt_tracker.completed ())
@@ -3617,7 +3617,7 @@ create_iso_definition (cpp_reader *pfile)
 	     function-like macros, but not at the beginning.  */
 	  if (macro->count == 1)
 	    {
-	      cpp_error (pfile, CPP_DL_ERROR, paste_op_error_msg);
+	      cpp_error (pfile, CPP_DL_ERROR, "%s", paste_op_error_msg);
 	      goto out;
 	    }
 
-- 
2.31.1

